// projectile/index.js
var VectorFactory = require("vector");
var oX = VectorFactory.CreateVector(0,0,1,0);

function Projectile() {
    this.damage = 1;
    this.type = "common-projectile";
    this.x = 0;
    this.y = 0;
    this.maxspeed = 15;
    this.speed = VectorFactory.CreateVector();
    this.markedForRemove = false;
    this.emitter = null;
}

Projectile.prototype.Move = function () {
    this.x += this.speed.x;
    this.y += this.speed.y;
};

Projectile.prototype.CalculateDirections = function () {
    this.Move();
    this.PreventGoingOutOfScreen();
};

Projectile.prototype.PreventGoingOutOfScreen = function () {
    if ((this.x < 0) || (this.y < 0) || (this.x > 1920) || (this.y > 1080))
        this.markedForRemove = true;
};

/* --- Projectile factory --- */

function ProjectileFactory() {}

ProjectileFactory.prototype.CreateProjectile = function (x,y, direction, emitter) {
    function spread() {
        return Math.random() * (1.2 - 0.8) + 0.8;
    }
    
    var projectile = new Projectile();
    projectile.x = x;
    projectile.y = y;
    projectile.speed = VectorFactory.CreateVector(0, 0, direction.x * spread(), direction.y * spread(), projectile.maxspeed);
    projectile.emitter = emitter;
    return projectile;
};

module.exports = new ProjectileFactory();
