// projectile/index.js
var VectorFactory = require("vector");
var oX = VectorFactory.CreateVector(0,0,1,0);

function Projectile() {
    this.id = Math.random();
    this.damage = 0;
    this.type = "";
    this.x = 0;
    this.y = 0;
    this.maxspeed = 0;
    this.speed = VectorFactory.CreateVector();
    this.markedForRemove = false;
    this.emitter = null;
}

Projectile.prototype.Move = function () {
    this.x += this.speed.x;
    this.y += this.speed.y;
};

Projectile.prototype.CalculateDirections = function () {
    if (this.emitter == null)
        return;
    this.Move();
    this.PreventGoingOutOfScreen();
};

Projectile.prototype.PreventGoingOutOfScreen = function () {
    if ((this.x < 0) || (this.y < 0) || (this.x > 1920) || (this.y > 1080))
        this.markedForRemove = true;
};

/* --- Projectile factory --- */

function ProjectileFactory() {}

ProjectileFactory.prototype.DetectStats = function (projectile) {
    switch (projectile.type) {
        default:
            projectile.maxspeed = 15;
            projectile.damage = 1;
    }
}

ProjectileFactory.prototype.CreateProjectile = function (x,y, direction, emitter, type) {
    var projectile = new Projectile();
    
    projectile.type = type || "common-projectile";    
    this.DetectStats(projectile);
    projectile.x = x;
    projectile.y = y;
    projectile.speed = VectorFactory.CreateVector(0, 0, direction.x, direction.y, projectile.maxspeed);
    projectile.emitter = emitter;
    
    return projectile;
};

module.exports = new ProjectileFactory();
