// mine/index.js
function Mine() {
    this.id = Math.random();
    this.damage = 0;
    this.type = "";
    this.x = 0;
    this.y = 0;
    this.markedForRemove = false;
    this.emitter = null;
    this.angle = 0;
    this.collisionRadius = 0;
    this.lifeSpan = 0;
}

Mine.prototype.CalculateDirections = function () {
    if (this.lifeSpan-- <= 0)
        this.markedForRemove = true;
};

/* --- Rocket factory --- */

function MineFactory() {}

MineFactory.prototype.DetectStats = function (projectile, timeScale) {
    switch (projectile.type) {
        default:
            projectile.lifeSpan = 1000;
            projectile.damage = 10;
            projectile.collisionRadius = 10;
            projectile.angle = Math.random() * 360;
    }
}

MineFactory.prototype.CreateMine = function (x,y, emitter, timeScale) {
    var projectile = new Mine();
    projectile.type = "common-mine-projectile";
    this.DetectStats(projectile, timeScale);
    projectile.x = x;
    projectile.y = y;
    projectile.emitter = emitter;
    return projectile;
};

module.exports = new MineFactory();
