// ship/index.js
var VectorFactory = require("vector");
var projectilePool = require("pool").Projectile;
var minePool = require("pool").Mine;
var rocketPool = require("pool").Rocket;

var oX = VectorFactory.CreateVector(0,0,1,0);
/* --- Ship --- */
function Ship() {
    this.hp = 0;
    this.id = Math.random();
    this.type = "common-ship";
    this.x = 0;
    this.y = 0;
    this.angle = 0;
    this.maxspeed = 0;
    this.acceleration = 0;
    this.speed = VectorFactory.CreateVector();
    this.state = "dead";
    this.respawnTimer = 0;
    this.name = "";
}

Ship.prototype.PreventGoingOutOfScreen = function () {
    if (this.x < 0)
        this.x = 0;
    if (this.y < 0)
        this.y = 0;
    if (this.x > 1920)
        this.x = 1920;
    if (this.y > 1080)
        this.y = 1080;
};

Ship.prototype.MoveShip = function () {
    this.x += this.speed.x;
    this.y += this.speed.y;
};

Ship.prototype.CalculateDirections = function (input) {
    var shipAcceleration = VectorFactory.CreateVector(this.x, this.y,
        input.x, input.y,
        this.acceleration);

    this.speed.Add(shipAcceleration);
    
    if (this.speed.magnitude > this.maxspeed)
        this.speed.Limit(this.maxspeed);
        
    this.angle = oX.AngleBetween(this.speed);
};

Ship.prototype.Damage = function (damage) {
    this.hp -= damage;
    
    if (this.hp < 1) {
        this.state = "dead";
        this.respawnTimer = 100;
    }
};

Ship.prototype.Fire = function () {
  projectilePool.ActivateProjectile(this.x, this.y, this.speed, this.id)
};

Ship.prototype.Rocket = function () {
  rocketPool.ActivateProjectile(this.x, this.y, this.speed, this.id)
};

Ship.prototype.Mine = function () {
  minePool.ActivateProjectile(this.x, this.y, this.id)
};

Ship.prototype.Respawn = function() {
    this.hp = 100;
    this.x = 1920 * Math.random();
    this.y = 1080 * Math.random();
    this.angle = 0;
    this.speed = VectorFactory.CreateVector();
    this.state = "alive";
};

/* --- Ship factory --- */

function ShipFactory() {}

ShipFactory.prototype.DetectStats = function(ship) {
    switch (ship.type) {
        default:
            ship.hp = 100;
            ship.maxspeed = 5;
            ship.acceleration = 0.3;
    }
};

ShipFactory.prototype.CreateShip = function(type) {
    var ship = new Ship();
    ship.type = type || "common-ship";
    this.DetectStats(ship);
    return ship;
};

/* --- Exports --- */
module.exports = new ShipFactory();
