// ship/index.js
var VectorFactory = require("vector");
var ProjectileFactory = require("projectile");
var oX = VectorFactory.CreateVector(0,0,1,0);
/* --- Ship --- */
function Ship() {
    this.type = "common-ship";
    this.x = 0;
    this.y = 0;
    this.angle = 0;
    this.maxspeed = 5;
    this.acceleration = 0.3;
    this.speed = VectorFactory.CreateVector()
}

Ship.prototype.PreventGoingOutOfScreen = function () {
    if (this.x < 0)
        this.x = 0;
    if (this.y < 0)
        this.y = 0;
    if (this.x > 1920)
        this.x = 1920;
    if (this.y > 1080)
        this.y = 1080;
};

Ship.prototype.MoveShip = function () {
    this.x += this.speed.x;
    this.y += this.speed.y;
};

Ship.prototype.CalculateDirections = function (input) {
    var shipAcceleration = VectorFactory.CreateVector(this.x, this.y,
        input.x, input.y,
        this.acceleration);

    this.speed.Add(shipAcceleration);
    
    if (this.speed.magnitude > this.maxspeed)
        this.speed.Limit(this.maxspeed);
        
    this.angle = oX.AngleBetween(this.speed);
};

//TODO : Instead of deleting implement using (getting porjectile) object pool
Ship.prototype.Fire = function (projectiles/*, pool*/) {
  projectiles.push(ProjectileFactory.CreateProjectile(this.x,this.y,this.speed));
};

/* --- Ship factory --- */

function ShipFactory() {}

ShipFactory.prototype.CreateShip = function(type) {
    var ship = new Ship();
    return ship;
};

/* --- Exports --- */
module.exports = new ShipFactory();
